proposals/0162-remove-accounts-executable-flag-checks.md

```diff
@@ -0,0 +1,137 @@
+type: Core
+status: Draft
+created: 2024-07-16
+feature: TBD
+extends: 0093
+---
+
+## Summary
+
+Remove all checks of accounts `is_executable` flag which can throw errors in
+the runtime.
+
+## Motivation
+
+Currently, a program account which satisfies the following conditions can be
+invoked / executed:
+
+- Has the `is_executable` flag set
+- Is owned by a loader
+- Contains a successfully verified ELF, contains the address of a program data
+account which does or is a built-in program
+
+The second condition can be used as a performance optimization to filter out
+accounts which definitely do not contain programs and the third condition alone
+would be sufficient to guarantee correct execution. The first condition is
+however is not only useless, it is downright deceptive and entangled with the
+confusing proxy account workaround in loader-v3.
+
+Originally, the `is_executable` flag meant that a program was deployed and
+finalized, which effectively renders its program account read-only forever,
+as the `is_executable` can not be cleared / reset.
+
+With the introduction of the upgradeable loader (loader-v3), instead of
+removing the first condition back then, a workaround was created: The program
+account became a proxy account which has the `is_executable` flag set and then
+points to the actual program data account which does not have the
+`is_executable` flag set. Meaning the `is_executable` flag now neither does
+reliably indicate that an account contains a program (as the program data
+account does not have the flag set), nor does it indicate that a program is
+finalized (as the program account has it set while the program remains
+upgradeable).
+
+Removing the first condition not only removes some redundant error checks from
+the runtime and confusion for future maintainers, but more importantly enables
+us to deploy loader-v4 which is upgradeable like loader-v3 but without the need
+for proxy accounts.
+
+## New Terminology
+
+None.
+
+## Detailed Design
+
+This proposal aims to unblock loader-v4 with minimal impact on the ecosystem by
+only removing checks of the `is_executable` flag. A complete removal of the
+flag can be addressed in a subsequent proposal. Thus, the following must remain
+unaffected for now:
+
+- Setting of the `is_executable` flag during program deployment
+- Calculation of the account hashes
+- Minimization of snapshots
+- Serialization of instruction accounts `is_executable` flag for dapps
+- CPI ignores changes made by the caller to instruction accounts which have
+the `is_executable` flag set
+
+These checks of the `is_executable` flag must be removed:
+
+- `ExecutableLamportChange` during execution (transaction succeeds instead)
+
+These checks of the `is_executable` flag are irrelevant to consensus as
+transactions which try to invoke accounts, which do not contain programs,
+continue to fail without them. They should still be removed because all
+implementations should aim to produce the same error codes:
```

Comment:
Might be worth mentioning that changing these error codes is non-consensus-breaking. It's implied here, but could be good to be specific.